File: typescript/matching-brackets/matching-brackets.ts
const MAP : {[k: string]: string} = {
  '[': ']',
  '(': ')',
  '{': '}'
}
export const isPaired = (str : string) => {
  const stk = []
  for (const n of str.split('')) {

    if (MAP[n]) {
      stk.push(n)
    } else if (Object.values(MAP).indexOf(n) !== -1) {
      if (!stk.length || MAP[stk.pop()!] !== n) {
        return false;
      }
    }

  }

  return stk.length === 0;
};

File: r/matching-brackets/matching-brackets.R
dict = list("{" = "}", "[" = "]", "(" = ")")

is_paired = function(s) {
  stk = list()
  for (c in s |> strsplit("") |> unlist()) {
    if (hasName(dict, c)) {
      stk = append(stk, dict[c])
    } else if (c %in% dict) {
      if (!length(stk)|| c != tail(stk, n = 1)) {
        return(FALSE)
      }
      stk = stk[-length(stk)]
    }
  }
  !length(stk)
}

